ha/core/controllers/pcs/node_controller.py

```diff
@@ -260,3 +314,70 @@ def shutdown(self, nodeid: str) -> dict:
+        check_cluster = op_kwargs.get("check_cluster") if op_kwargs.get("check_cluster") is not None else True
+        poweroff = op_kwargs.get("poweroff") if op_kwargs.get("poweroff") is not None else False
+        storageoff = op_kwargs.get("storageoff") if op_kwargs.get("storageoff") is not None else False
+        try:
+            # Raise exception if node_id is not valid
+            self._is_node_in_cluster(node_id=node_id)
+
+            # Get the nodeid from pvtfqdn
+            nodeid = self._health_manager.get_key(f"{const.PVTFQDN_TO_NODEID_KEY}/{node_id}")
+
+            # stop all services running on node with node_id
+            node_status = self._system_health.get_node_status(node_id=nodeid).get("status")
+            if node_status == HEALTH_STATUSES.OFFLINE.value:
+                Log.info(f"For stop {node_id}, Node already in offline state.")
+                status = f"Node {node_id} is already in offline state."
+                return {"status": const.STATUSES.SUCCEEDED.value, "output": node_status, "error": "", "msg": status}
+            elif node_status == HEALTH_STATUSES.FAILED.value:
+                raise ClusterManagerError(f"Failed to stop {node_id}. node is in {node_status} state.")
+            else:
+                if check_cluster:
+                    # Checks whether cluster is going to be offline if node with node_id is stopped.
+                    res = json.loads(self.check_cluster_feasibility(node_id=node_id))
+                    if res.get("status") == const.STATUSES.FAILED.value:
+                        return res
+                if self.heal_resource(node_id):
+                    time.sleep(const.BASE_WAIT_TIME)
+
+            if storageoff:
+                # Stop services on node except sspl-ll
+                self._controllers[const.SERVICE_CONTROLLER].stop(node_id=node_id, excludeResourceList=[RESOURCE.SSPL_LL.value])
+
+                # TODO: storage enclosure is stopped on the node
+
+                # Put node in standby mode
+                self._execute.run_cmd(const.PCS_NODE_STANDBY.replace("<node>", node_id), f" --wait={const.CLUSTER_STANDBY_UNSTANDBY_TIMEOUT}")
+                Log.info(f"Executed node standby for {node_id}")
+                self._controllers[const.SERVICE_CONTROLLER].clear_resources(node_id=node_id)
+            else:
+                self._execute.run_cmd(const.PCS_NODE_STANDBY.replace("<node>", node_id), f" --wait={const.CLUSTER_STANDBY_UNSTANDBY_TIMEOUT}")
+                Log.info(f"Executed node standby for {node_id}")
+            status = f"{node_id} Node Standby is in progress"
+
+            # Update node health
+            initial_event = self._system_health.get_health_event_template(nodeid=nodeid, event_type=HEALTH_EVENTS.FAULT.value)
+            Log.debug(f"Node health : {initial_event} updated for node {node_id}")
+            health_event = HealthEvent.dict_to_object(initial_event)
+            self._system_health.process_event(health_event)
```

Comment:
Do we need to update node health for VM also

Reply:
Please add a TODO in VM code to update node health

Reply:
Done, added TODO